src/apiserver/interface_provider.cc

```diff
@@ -0,0 +1,197 @@
+//     copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+//     copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "apiserver/interface_provider.h"
+
+#include <deque>
+#include <regex>  // NOLINT
+#include <stdexcept>
+
+#include "boost/algorithm/string/split.hpp"
+
+namespace fedb {
+namespace http {
+
+std::vector<std::unique_ptr<PathPart>> Url::parsePath(bool disableIds) const {
+    std::deque<std::string> splitted;
+    boost::algorithm::split(splitted, path, [](char c) { return c == '/'; });
+    splitted.pop_front();
+
+    std::vector<std::unique_ptr<PathPart>> splitPath;
+    for (auto const& i : splitted) {
+        if (!disableIds && i.front() == ':')
+            splitPath.emplace_back(new PathParameter(i.substr(1, i.length() - 1)));
+        else
+            splitPath.emplace_back(new PathString(i));
+    }
+    return splitPath;
+}
+
+PathParameter::PathParameter(std::string id) : value_(), id_(id) {}
+
+std::string PathParameter::getValue() const { return value_; }
+
+std::string PathParameter::getId() const { return id_; }
+
+void PathParameter::setValue(std::string const& value) { value_ = value; }
+
+PathType PathParameter::getType() const { return PathType::PARAMETER; }
+
+PathString::PathString(std::string value) : value_(value) {}
+
+std::string PathString::getValue() const { return value_; }
+
+PathType PathString::getType() const { return PathType::STRING; }
+
+void ReducedUrlParser::parseQuery(std::string const& query, Url* url) {
+    std::regex rgx(R"((\w+=(?:[\w-])+)(?:(?:&|;)(\w+=(?:[\w-])+))*)");
+    std::smatch match;
+
+    if (std::regex_match(query, match, rgx)) {
+        for (auto i = std::begin(match) + 1; i < std::end(match); ++i) {
+            auto pos = i->str().find_first_of('=');
+            url->query[i->str().substr(pos + 1)] = i->str().substr(0, pos);
+        }
+    }
+}
+
+Url ReducedUrlParser::parse(std::string const& urlString) {
+    Url url;
+    url.url = urlString;
+
+    // regex for extracting path, query, fragment
+    // (?:(?:(\/(?:(?:[a-zA-Z0-9]|[-_~!$&']|[()]|[*+,;=:@])+(?:\/(?:[a-zA-Z0-9]|[-_~!$&']|[()]|[*+,;=:@])+)*)?)|\/)?
+    // (?:(\?(?:\w+=(?:[\w-])+)(?:(?:&|;)(?:\w+=(?:[\w-])+))*))?(?:(#(?:\w|\d|=|\(|\)|\\|\/|:|,|&|\?)+))?)
+    std::regex rgx(

```

Comment:
Is there a perfermance issue if we initialize `regex` when using it.

Reply:
I'll test it.

Reply:
--------------------------------------------------------------
Benchmark                    Time             CPU   Iterations
--------------------------------------------------------------
BM_UrlParse             817332 ns       817312 ns          696
BM_UrlParseWithInit        787 ns          787 ns       896177

It's better to init `std::regex` first, instead of creating it in every `parse`. I'll fix it.